### PR TITLE
NFT setApprovalForAll  warning popover

### LINKED ISSUE #15700: Create "your NFTs may be at risk" modal
## Resources
[Figma Design file](https://www.figma.com/file/n0xMnZqQz4w9Sfa3ZdhWUi/Consentful-experience-for-token-allowances?node-id=2007%3A14561)

## Acceptance Criteria
1. Total is the amount of NFTs the user currently owns from that collection;
2. The existing account component should be used;
3. The copy displayed reads: `You're granting access to **all your {NFTCollectionName} NFTs**, including any you might own in the future. The party on the other end can transfer these NFTs from your wallet at any time without asking you until you revoke this approval. **Proceed with caution.**` (Use the wording from AC and not from design)
4. Clicking on `Reject` rejects the transaction request (same behavior as clicking on "Reject" on the previous screen or in the current existing screen); (Use 'Reject" instead of Cancel from design)
5. Clicking on `Approve` submits the transaction (same behavior as clicking on Approve in the current existing screen).
Screen Shot 2022-08-24 at 19 37 11

### LINKED ISSUE #15701: Update "Approve" button behavior on SetApprovalForAll confirmation screens
## Resources
[Figma Design file](https://www.figma.com/file/n0xMnZqQz4w9Sfa3ZdhWUi/Consentful-experience-for-token-allowances?node-id=2007%3A14561)

## Acceptance Criteria
1. Clicking on "Approve" opens the "your NFTs may be at risk" modal that will be created on https://github.com/MetaMask/metamask-extension/issues/15700

## Steps to Reproduce
1. Open the test dapp https://metamask.github.io/test-dapp/
4. Under the collectibles section, click on "Deploy"
5. Wait for transaction to be confirmed and click on "Mint"
6. Wait for transaction to be confirmed and click on "Set Approval For All"

### PR DESCRIPTION
**Pull request by amerkadicE**
## Explanation

This PR will add a new warning popover modal when the user approves the collection of NFTs.

NOTE: We will use a center justified modal, Figma design is not updated.
https://github.com/MetaMask/metamask-extension/issues/15700#issuecomment-1277716840

## More Information

Fixes #15700 
Fixes #15701 

### Before

It is a new modal.

### After

https://user-images.githubusercontent.com/97883527/195564521-9bd00b43-5674-405d-99f8-4d1a2b9cfdfe.mov

## Manual Testing Steps

Open the test dapp https://metamask.github.io/test-dapp/
Under the collectibles section, click on "Deploy"
Wait for transaction to be confirmed and click on "Mint"
Wait for transaction to be confirmed and click on "Set Approval For All"
Click Approve button
Notice there is a new nft warning popover modal displayed

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [ ] PR has been added to the appropriate release Milestone

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [ ] "Extension QA Board" label has been applied

### CODE REVIEW

**Comment by amerkadicE** [comment]
> This looks great! Thanks, @amerkadicE ! Small additional ask: could we also remove the warning (screenshot below) in this PR? Screen Shot 2022-10-13 at 09 38 54

Done

File: ui/components/app/confirm-page-container/confirm-page-container.component.js
Comment (danjm) [inline ui/components/app/confirm-page-container/confirm-page-container.component.js:123]: Do we know if `tokenBalance = await fetchTokenBalance(tokenAddress, fromAddress); collectionBalance = tokenBalance?.balance?.words?.[0] || 0;` will work for all NFT contracts that have a setApprovalForAll method?

Maybe @adonesky1 knows?

File: ui/components/app/confirm-page-container/confirm-page-container-warning/confirm-page-container-warning.js
Comment (danjm) [inline ui/components/app/confirm-page-container/confirm-page-container-warning/confirm-page-container-warning.js:1]: Is this component only going to be used for `setApprovalForAll` confirmations? If so, we should name the component and the file in a way that communicates that (e.g. `set-approval-for-all-warning.js`) and probably move it to `ui/components/app`

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (danjm) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:1233]: Because this is not used for all token approvals, the code would be clearer if this was renamed to indicate that it is only used for `setApprovalForAll` (e.g. `onSetApprovalForAll` and `handleSetApprovalForAll`)

**Review by danjm — DISMISSED** [review]
I've left some comments on improving the readability of the code, and one on a performance improvement. Also one question we should confirm an answer to before we merge

File: ui/components/app/confirm-page-container/confirm-page-container.component.js
Comment (amerkadicE) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js]: Done

File: ui/components/app/confirm-page-container/confirm-page-container-warning/confirm-page-container-warning.js
Comment (amerkadicE) [reply ui/components/app/confirm-page-container/confirm-page-container-warning/confirm-page-container-warning.js]: Done

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (amerkadicE) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js]: Done

File: app/_locales/en/messages.json
Comment (georgewrmarshall) [inline app/_locales/en/messages.json:2254]: I'm not sure what the best practices for content are but should there be a description explaining what `$1` and `$2` are similarly to `nftWarningContentBold`?

File: ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js
Comment (georgewrmarshall) [inline ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js:121]: Would be nice to get some propType descriptions here

**Review by georgewrmarshall — COMMENTED** [review]
Looking good! Would be nice to get a story for `SetApproveForAllWarning`

File: app/_locales/en/messages.json
Comment (amerkadicE) [reply app/_locales/en/messages.json]: done

File: ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js
Comment (amerkadicE) [reply ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js]: done

File: ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.stories.js
Comment (georgewrmarshall) [inline ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.stories.js:27]: Nice! You can add the argTypes and default args to the storybook options as well as add the action controls

```suggestion
  argTypes: {
    collectionName: {
      control: 'text',
    },
    name: {
      control: 'text',
    },
    senderAddress: {
      control: 'text',
    },
    total: {
      control: 'text',
    },
    onSubmit: {
      action: 'onSubmit',
    },
    onCancel: {
      action: 'onCancel',
    },
  },
  args: {
    collectionName: 'Test collection',
    name: 'Account 1',
    senderAddress: '0xee014609ef9e09776ac5fe00bdbfef57bcdefebb',
    total: '6',
  },
};

export const DefaultStory = (args) => <SetApproveForAllWarning {...args} />;

DefaultStory.storyName = 'Default';
```

**Review by georgewrmarshall — COMMENTED** [review]
UI LGTM! Have a small suggestion for the story

**Review by georgewrmarshall — DISMISSED** [review]
lgtm pending @danjm's suggestions
- checked story
- checked manual testing steps

**Comment by danjm** [comment]
I think the last thing before we merge this is to confirm the answer to this question `https://github.com/MetaMask/metamask-extension/pull/16195#discussion_r999700868`

In case somehow there are types of contracts that require different methods of getting the balance

@adonesky1 any thoughts here?

File: ui/components/app/confirm-page-container/confirm-page-container.component.js
Comment (adonesky1) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js]: @danjm you are correct that this won't work for all NFT contracts. `fetchTokenBalance` uses the `balanceOf` method, which on ERC721's only requires the address to check against as an arg, [but for ERC1155's requires](https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#IERC1155-balanceOf-address-uint256-) the both the `address` and the `tokenID`. This is because 1155's have the concept of fungible sets of tokenIds so there could be 100 tokens on the contract with `tokenId === 1` and 1000 with `tokenId === 2` all of which are interchangeable. So unfortunately we can't be sure of a users' balance simply on a given ERC1155 contract simply querying directly against the contract (unless we already know all of the tokenIDs they own). We can still achieve this for ERC721 contracts. But for 1155's we'll have to use an indexer like OpenSea to know user balances.

Comment (bschorchit) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js]: My thinking was that we could add in a different ticket/PR the logic here to check if OpenSea API is enabled and use it for getting the balance so we don't block this until we're ready to move the OpenSea API part from behind the NFT feature flag.

Comment (adonesky1) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js]: Sounds right to me!

Comment (amerkadicE) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js]: @bschorchit Sure, I will update my PR to show the Total number only for ERC721 standard

Comment (amerkadicE) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js]: @bschorchit @adonesky1 
Done, now it will show the total number if the contract follows the ERC721 standard.
You can check it in the storybook also.

**Comment by bschorchit** [comment]
> I think the last thing before we merge this is to confirm the answer to this question `https://github.com/MetaMask/metamask-extension/pull/16195#discussion_r999700868`
> 
> In case somehow there are types of contracts that require different methods of getting the balance
> 
> @adonesky1 any thoughts here?

Great catch, Dan! This has been addressed [here](https://github.com/MetaMask/metamask-extension/pull/16195#discussion_r1020113641) and is now ready for a new review :)

**Comment by mirjanaKukic** [comment]
Verified by QA

**Comment by bschorchit** [comment]
Ok, so turns out the NFT name is not always available and neither is its symbol. 
We should pull out [the existing logic here](https://github.com/MetaMask/metamask-extension/blob/develop/ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js#L453) to a helper function and use it to determine whether to use name, symbol or something else here in this modal. cc: @amerkadicE

File: ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js
Comment (adonesky1) [inline ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js:84]: [nit] ?
```suggestion
        {isERC721 && total <Typography>{`${t('total')}: ${total}`}</Typography>}
```

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (adonesky1) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:1234]: I'm not too worried about this since I'd imagine we're going to rebuild these confirm transaction components soon enough, but should we abstract this away from the specific case of `setApprovalForAll` so that we could potentially show warning modals with different content for different kinds of transactions?

File: ui/components/app/confirm-page-container/confirm-page-container.component.js
Comment (adonesky1) [inline ui/components/app/confirm-page-container/confirm-page-container.component.js:376]: We should delete this whole block... This appears to just leave a weird piece of dead/unreachable code

Comment (amerkadicE) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js]: Removed

File: ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js
Comment (amerkadicE) [reply ui/components/app/set-approval-for-all-warning/set-approval-for-all-warning.js]: Resolved

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (brad-decker) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:1234]: We should not consider refactoring this at this point, it should be considered holistically separate from (but including) this change. 

> maybe its okay

okay is probably a strong word to use here. It is the least bad option to just extend this pattern of overloading props etc, rather than trying to find a way to abstract out things without considering the entirety of the prop overloading problem and organization of confirmations. I do have opinions on confirmation restructuring as a whole, in that I thought we were going to try and use the template system for them -- but perhaps that is old thinking. It has been a **long** time since the first confirmations were introduced to use that system.  The idea there was that using the template system allowed for snaps authors to interweave their own UI into ours. We should probably come to consensus on the path forward there.

Comment (adonesky1) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:1234]: Ok sounds goood!

**Review by adonesky1 — APPROVED** [review]
LGTM!

File: ui/components/app/confirm-page-container/confirm-page-container.component.js
Comment (brad-decker) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:120]: @pedronfigueiredo, @bschorchit @amerkadicE -- Love this suggestion considering this component is low complexity. I'm in favor of either

A) blocking PR until this is completed as we are introducing a componentDidMount here which actually **adds** complexity for a future PR to convert this to functional.
B) Freeing @amerkadicE up to do this work in a PR immediately to follow this one. 

Either option helps to pay off some debt instead of incurring new debt.

Comment (bschorchit) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:120]: > @pedronfigueiredo, @bschorchit @amerkadicE -- Love this suggestion considering this component is low complexity. I'm in favor of either
> 
> 
> 
> A) blocking PR until this is completed as we are introducing a componentDidMount here which actually **adds** complexity for a future PR to convert this to functional.
> 
> B) Freeing @amerkadicE up to do this work in a PR immediately to follow this one. 
> 
> 
> 
> Either option helps to pay off some debt instead of incurring new debt. 

I vote B so the user facing changes (added friction for this request) reach the users sooner :) 
I'm happy to create the ticket for the refactoring so it gets worked on next.

Comment (adonesky1) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:120]: I agree I think option B works!

**Review by brad-decker — APPROVED** [review]

File: ui/components/app/confirm-page-container/confirm-page-container.component.js
Comment (bschorchit) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:120]: For visibility, new ticket created here: https://github.com/MetaMask/metamask-extension/issues/16824

Comment (amerkadicE) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:120]: You can assign it to me @bschorchit

Comment (pedronfigueiredo) [reply ui/components/app/confirm-page-container/confirm-page-container.component.js:120]: great, thanks all

**Pull request merged**
